object_store/src/integration.rs

```diff
@@ -1196,3 +1196,33 @@ pub async fn multipart_race_condition(storage: &dyn ObjectStore, last_writer_win
         ));
     }
 }
+
+/// Tests performing out of order multipart uploads
+pub async fn multipart_out_of_order(storage: &dyn ObjectStore) {
+    let path = Path::from("test_multipart_out_of_order");
+    let mut multipart_upload = storage.put_multipart(&path).await.unwrap();
+
+    let part1 = std::iter::repeat(b'1')
+        .take(5 * 1024 * 1024)
+        .collect::<Bytes>();
+    let part2 = std::iter::repeat(b'2')
+        .take(5 * 1024 * 1024)
+        .collect::<Bytes>();
+    let part3 = std::iter::repeat(b'3')
+        .take(5 * 1024 * 1024)
+        .collect::<Bytes>();
+    let full = [part1.as_ref(), part2.as_ref(), part3.as_ref()].concat();
+
+    let fut1 = multipart_upload.put_part(part1.into());
+    let fut2 = multipart_upload.put_part(part2.into());
+    let fut3 = multipart_upload.put_part(part3.into());
+    fut2.await.unwrap();
```

Comment:
I missed the out of order part of the first read of this PR. Maybe a comment would help:
```suggestion
    // note order is 2,3,1 , different than the parts were created in
    fut2.await.unwrap();
```